docs/ovs-pipeline.md

```diff
@@ -0,0 +1,561 @@
+
+### ConntrackStateTable (31)
+
+This table handles all "tracked" packets (all packets are moved to the tracked
+state by the previous table, [ConntrackTable]). It serves the following
+purposes:
+ * keeps track of connections going through the gateway port; for all packets
+ belonging to such connections we overwrite the destination MAC to the local
+ gateway MAC on their way to the gateway port. This is required to handle the
+ case were a local Pod connects to a service for which the selected backend is
+ also a local Pod. In this specific case, if we do not use connection tracking,
+ response traffic from the backend Pod will go directly to the originating Pod
+ (L2-switched) without going first through the gateway and kube-proxy. This
+ means that the response traffic will arrive at the originating Pod with the
+ incorrect source IP (it will be set to the Pod backend's IP instead of the
+ service IP). So we keep track of such connections (indentified by matching the
+ ingress port against the gateway port) and for all response traffic we set the
+ destination to the local gateway MAC.
+ * commit all new connections
+ * drop packets reported as invalid by conntrack
+
+If you dump the flows for this table, you should see the following:
+```
+1. table=31, priority=210,ct_state=-new+trk,ct_mark=0x20,ip,reg0=0x1/0xffff actions=resubmit(,40)
+2. table=31, priority=200,ct_state=+new+trk,ip,reg0=0x1/0xffff actions=ct(commit,table=40,zone=65520,exec(load:0x20->NXM_NX_CT_MARK[],move:NXM_OF_ETH_SRC[]->NXM_NX_CT_LABEL[0..47]))
+3. table=31, priority=200,ct_state=-new+trk,ct_mark=0x20,ip actions=move:NXM_NX_CT_LABEL[0..47]->NXM_OF_ETH_DST[],resubmit(,40)
+4. table=31, priority=200,ct_state=+inv+trk,ip actions=drop
+5. table=31, priority=190,ct_state=+new+trk,ip actions=ct(commit,table=40,zone=65520)
+6. table=31, priority=80,ip actions=resubmit(,40)
+```
+
+Flows 1,2 and 3 implement the destination MAC rewrite described above. Note that
+at the moment we use the lower 48 bits of NXM_NX_CT_LABEL to cache the MAC
+address and we use the cached value for the rewrite. This is a bit of an
+overkill since this cached value is always the MAC address of the local
+gateway, which is a known, static value. In the future we may get rid of
+NXM_NX_CT_LABEL altogether and use a `load` action instead of a `move` action
+for the rewrite.
+
+Flow 5 commits all connections to the conntrack module.
+
+All traffic is finally resubmitted to the [DnatTable].
+
+### DnatTable (40)
+
+At the moment this table's only job is to send traffic destined to services
+through the local gateway, without any modifications. kube-proxy will then take
+care of load-balancing the connections across the different Pod backends for
+each service.
+
+If you dump the flows for this table, you should see something like this:
+```
+1. table=40, priority=200,ip,nw_dst=10.96.0.0/12 actions=output:gw0
+2. table=40, priority=80,ip actions=resubmit(,50)
```

Comment:
priority=80 will be replaced by priority=0 in all tables once PR: #195 is merged.